src/main/java/uk/gov/hmcts/reform/waworkflowapi/clients/service/ExternalTaskWorker.java

```diff
@@ -48,20 +52,49 @@ public void setupClient() {
     }
 
     public void checkIdempotency(ExternalTask externalTask, ExternalTaskService externalTaskService) {
-        Boolean isDuplicate =  externalTask.getVariable("isDuplicate");
-        if (isDuplicate == null || isDuplicate) {
-            Map<String, Object> processVariables = singletonMap(
-                "isDuplicate",
-                false
-            );
+        IdempotentId idempotentId = getIdempotentId(externalTask);
 
-            LOGGER.info("Duplicate was hit.");
+        Optional<IdempotentKeys> idempotentRow = idempotentKeysRepository.findById(idempotentId);
 
-            externalTaskService.complete(externalTask, processVariables);
-        } else {
-            externalTaskService.complete(externalTask);
-            LOGGER.info("Duplicate was true");
+        idempotentRow.ifPresentOrElse(
+            (row) -> handleIdempotentIdIsPresentInDb(externalTask, externalTaskService, row),
+            () -> handleIdempotentIdIsNotPresentInDb(externalTask, externalTaskService, idempotentId)
+        );
+    }
 
+    private IdempotentId getIdempotentId(ExternalTask externalTask) {
+        String idempotentKey = externalTask.getVariable("idempotentKey");
```

Comment:
I think we need to handle the case where no idempotentKey is received e.g. null or empty string.  As above it's a common bpmn that other teams could use.  We can decide that a duplicate check is not required in this case and respond as if no duplicate is found.  We should probably log that decision in the code clearly though.  Open to suggestion for alternatives.

Reply:
Let's discuss tomorrow.

Reply:
to address as part of the new ticket?

Reply:
As discussed in Planning we can do this in the new unhappy paths ticket.  But let's make sure it's not missed.

Reply:
Done. Covered by unit test only.